### PR TITLE
[IN PROGRESS] Chunk refactor

### PR DESCRIPTION
**Pull request by gbalint**

### CODE REVIEW

File: swarm/network/stream/common_test.go [hunk 1]
```diff
@@ -102,32 +105,56 @@ func NewStreamerService(ctx *adapters.ServiceContext) (node.Service, error) {
 	var err error
```
Comment (holisticode) [inline swarm/network/stream/common_test.go:104]: The changes in this file suggest it's best to wait for this PR to be merged before we refactor the simulations....

File: swarm/network/stream/delivery.go [hunk 2]
```diff
@@ -42,17 +43,17 @@ var (
 )
 
 type Delivery struct {
```
Comment (holisticode) [inline swarm/network/stream/delivery.go:45]: I think `Delivery` is seriously lacking documentation, also how the different parts play together - `Delivery`, SwarmChunkServer`, `SwarmChunkClient`, `ChunkStore`, etc.

I know it is a bit of a debt from the previous implementation but it is a good moment to add it.

File: swarm/network/stream/delivery_test.go [hunk 3]
```diff
@@ -234,16 +247,16 @@ func TestStreamerDownstreamChunkDeliveryMsgExchange(t *testing.T) {
...
-
-	if !created {
-		t.Fatal("chunk already exists")
-	}
-	select {
-	case <-chunk.ReqC:
-		t.Fatal("chunk is already received")
-	default:
-	}
+	// chunk, created := localStore.GetOrCreateRequest(chunkKey)
```
Comment (holisticode) [inline swarm/network/stream/delivery_test.go]: why is all this code commented?

File: swarm/network/stream/delivery_test.go [hunk 4]
```diff
@@ -310,7 +317,7 @@ func TestDeliveryFromNodes(t *testing.T) {
 func testDeliveryFromNodes(t *testing.T, nodes, conns, chunkCount int, skipCheck bool) {
 	defaultSkipCheck = skipCheck
 	toAddr = network.NewAddrFromNodeID
-	createStoreFunc = createTestLocalStorageFromSim
+	// createStoreFunc = createTestLocalStorageFromSim
```
Comment (holisticode) [inline swarm/network/stream/delivery_test.go:316]: remove if not needed

File: swarm/network/stream/messages.go [hunk 5]
```diff
@@ -194,15 +197,17 @@ func (p *Peer) handleOfferedHashesMsg(req *OfferedHashesMsg) error {
 		return fmt.Errorf("error initiaising bitvector of length %v: %v", len(hashes)/HashSize, err)
 	}
 	wg := sync.WaitGroup{}
+	ctx, cancel := context.WithTimeout(context.Background(), syncBatchTimeout)
```
Comment (holisticode) [inline swarm/network/stream/messages.go]: no defer cancel?

File: swarm/network/stream/snapshot_retrieval_test.go [hunk 6]
```diff
@@ -80,6 +80,7 @@ func initRetrievalTest() {
 //Files are uploaded to nodes, other nodes try to retrieve the file
 //Number of nodes can be provided via commandline too.
 func TestFileRetrieval(t *testing.T) {
+	t.Skip("not working")
```
Comment (holisticode) [inline swarm/network/stream/snapshot_retrieval_test.go:83]: Add a comment why the tests are disabled please

File: swarm/network/stream/snapshot_sync_test.go [hunk 7]
```diff
@@ -117,6 +120,7 @@ func initSyncTest() {
 //they are expected to store based on the syncing protocol.
 //Number of chunks and nodes can be provided via commandline too.
 func TestSyncing(t *testing.T) {
+	t.Skip("not working")
```
Comment (holisticode) [inline swarm/network/stream/snapshot_sync_test.go]: ????
Why is this test disabled???

File: swarm/network/stream/stream.go [hunk 8]
```diff
@@ -323,9 +322,9 @@ func (r *Registry) Quit(peerId discover.NodeID, s Stream) error {
 	return peer.Send(msg)
 }
 
-func (r *Registry) Retrieve(chunk *storage.Chunk) error {
-	return r.delivery.RequestFromPeers(chunk.Addr[:], r.skipCheck)
-}
+// func (r *Registry) Retrieve(ctx context.Context, chunk storage.Chunk) (context.Context, error) {
+// 	return r.delivery.RequestFromPeers(ctx, chunk.Address()[:], nil, r.skipCheck, &sync.Map{})
+// }
```
Comment (holisticode) [inline swarm/network/stream/stream.go:328]: commented code

File: swarm/network/stream/syncer.go [hunk 9]
```diff
@@ -27,45 +28,55 @@ import (
 )
 
 const (
-	// BatchSize = 2
 	BatchSize = 128
 )
 
+type SyncDB interface {
```
Comment (holisticode) [inline swarm/network/stream/syncer.go]: Please some more documentation, what are these interfaces for

File: swarm/network/stream/syncer.go [hunk 10]
```diff
@@ -203,26 +210,21 @@ func NewSwarmSyncerClient(p *Peer, db *storage.DBAPI, ignoreExistingRequest bool
...
-	if chunk.ReqC == nil { //|| (s.ignoreExistingRequest && !created) {
-		return nil
-	}
-	// create request and wait until the chunk data arrives and is stored
-	return func() {
-		chunk.WaitToStore()
-	}
+func (s *SwarmSyncerClient) NeedData(ctx context.Context, key []byte) (wait func(context.Context) error) {
+	fetch := s.store.(Has).Has(ctx, key)
+	// return func(ctx context.Context) error {
```
Comment (holisticode) [inline swarm/network/stream/syncer.go]: What's this commented code

File: swarm/storage/chunker.go [hunk 11]
```diff
@@ -370,48 +365,63 @@ func (tc *TreeChunker) Append() (Address, func(), error) {
...
 		}
-		if chunkData == nil {
-			select {
-			case <-quitC:
-				return 0, errors.New("aborted")
-			default:
-				return 0, fmt.Errorf("root chunk not found for %v", r.key.Hex())
-			}
-		}
+		// when can this happen? I commenting this out there should be an error upstream
```
Comment (holisticode) [inline swarm/storage/chunker.go:402]: commented code

File: swarm/storage/chunker.go [hunk 12]
```diff
@@ -370,48 +365,63 @@ func (tc *TreeChunker) Append() (Address, func(), error) {
...
+		// 	default:
+		// 		return 0, fmt.Errorf("root chunk not found for %v", r.key.Hex())
+		// 	}
+		// }
 		r.chunkData = chunkData
+		s := r.chunkData.Size()
+		log.Debug("lazychunkreader.size", "key", r.addr, "size", s)
+		if s < 0 {
+			panic("corrupt size")
+			return 0, errors.New("corrupt size")
```
Comment (holisticode) [inline swarm/storage/chunker.go]: ?

Code after `panic` should not run?

File: swarm/storage/filestore_test.go [hunk 13]
```diff
@@ -48,31 +51,39 @@ func testFileStoreRandom(toEncrypt bool, t *testing.T) {
...
-	reader, slice := generateRandomData(testDataSize)
-	key, wait, err := fileStore.Store(reader, testDataSize, toEncrypt)
+	reader, slice := GenerateRandomData(testDataSize)
+	ctx, cancel := context.WithTimeout(context.Background(), splitTimeout)
+	defer cancel()
+	key, wait, err := fileStore.Store(ctx, reader, testDataSize, toEncrypt)
+	if err != nil {
+		t.Fatalf("Store error: %v", err)
+	}
+	log.Warn("store called")
```
Comment (holisticode) [inline swarm/storage/filestore_test.go]: Why is this a warn

File: swarm/storage/filestore_test.go [hunk 14]
```diff
@@ -48,31 +51,39 @@ func testFileStoreRandom(toEncrypt bool, t *testing.T) {
...
+	err = wait(ctx)
 	if err != nil {
-		t.Errorf("Store error: %v", err)
+		t.Fatalf("Store error: %v", err)
 	}
-	wait()
-	resultReader, isEncrypted := fileStore.Retrieve(key)
+	log.Warn("store complete", "address", key.Hex())
```
Comment (holisticode) [inline swarm/storage/filestore_test.go]: probably shouldn't be warn either

File: swarm/storage/ldbstore.go [hunk 15]
```diff
@@ -199,13 +215,14 @@ func getIndexKey(hash Address) []byte {
...
 }
 
-func getOldDataKey(idx uint64) []byte {
-	key := make([]byte, 9)
-	key[0] = keyOldData
-	binary.BigEndian.PutUint64(key[1:9], idx)
-
-	return key
-}
+//
```
Comment (holisticode) [inline swarm/storage/ldbstore.go:218]: commented code

File: swarm/storage/ldbstore.go [hunk 16]
```diff
@@ -221,28 +238,28 @@ func encodeIndex(index *dpaDBIndex) []byte {
...
-	chunk.Size = int64(binary.BigEndian.Uint64(data[32:40]))
+func decodeData(addr Address, data []byte) (*chunk, error) {
+	return NewChunk(addr, data[32:]), nil
 }
 
-func decodeOldData(data []byte, chunk *Chunk) {
-	chunk.SData = data
-	chunk.Size = int64(binary.BigEndian.Uint64(data[0:8]))
-}
+//
```
Comment (holisticode) [inline swarm/storage/ldbstore.go:258]: commented code

File: swarm/storage/localstore_test.go [hunk 17]
```diff
@@ -30,6 +30,7 @@ var (
 // tests that resource update chunks are passed through content address validator
 // the test checking the resouce update validator internal correctness is found in resource_test.go
 func TestValidator(t *testing.T) {
+	t.Skip("just for now")
```
Comment (holisticode) [inline swarm/storage/localstore_test.go]: Is it still for now or did this get forgotten

**Review by holisticode — CHANGES_REQUESTED** [review]
This is a huge PR, and after reviewing intensively a first part I realized I didn't have the concentration anymore to review the rest.

One suggestion could be to split up so we can review different parts

**Comment by gbalint** [comment]
@holisticode This PR is totally not ready for review, it just exists as a shared branch between @zelig and me so we can cooperate. I see that you received an automatic review request, that is just a fuckup in the github settings, sorry for that

File: swarm/network/fetcher.go [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+		skipCheck: skipCheck,
+	}
+}
+
+// fetch is called by NetStore evey time there is a request or offer for a chunk
+func (f *Fetcher) fetch(ctx context.Context) {
+	// put source/request
+	var source *discover.NodeID
+	if sourceIF := ctx.Value("source"); sourceIF != nil {
+		id := discover.MustHexID(sourceIF.(string))
```
[the remark this answers is not in the document]
Comment (zelig) [reply swarm/network/fetcher.go:94]: to take `&` you need to assign

File: swarm/network/fetcher.go [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+		source = &id
+	}
+	select {
+	case f.sourceC <- source:
+	case <-ctx.Done():
+	}
+}
+
+// start prepares the Fetcher
```
[the remark this answers is not in the document]
Comment (zelig) [reply swarm/network/fetcher.go:103]: y

File: swarm/network/fetcher.go [hunk 20]
```diff
@@ -0,0 +1,239 @@
...
+	// rerequesting extends the search. ie.,
+	// if a peer we requested from is gone we issue a new request, so the number of active
+	// requests never decreases
+	for {
+		select {
+
+		// accept a request or source.
+		case source := <-f.sourceC:
+			if source != nil {
+				log.Debug("new source", "peer addr", source, "request addr", f.addr)
```
[the remark this answers is not in the document]
Comment (zelig) [reply swarm/network/fetcher.go:127]: the same thing

File: swarm/network/fetcher.go [hunk 21]
```diff
@@ -0,0 +1,239 @@
...
+				// 1) the chunk is offered by a syncing peer
+				// add to known sources
+				requested = true
+				sources = append(sources, source)
```
[the remark this answers is not in the document]
Comment (zelig) [reply swarm/network/fetcher.go]: ? its a local variable only changed in the loop

File: swarm/network/fetcher.go [hunk 22]
```diff
@@ -0,0 +1,239 @@
...
+				// launch a request to the source iff the chunk was requested (not just expected because its offered by a syncing peer)
+				doRequest = requested
+			} else {
+				log.Debug("new request", "request addr", f.addr)
```
[the remark this answers is not in the document]
Comment (zelig) [reply swarm/network/fetcher.go:134]: if `source == nil` it is a genuine request not an offer from an upstream peer syncing

File: swarm/network/fetcher.go [hunk 23]
```diff
@@ -0,0 +1,239 @@
...
+				// 2) chunk is requested, set requested flag
+				// launch a request iff none been launched yet
+				doRequest = !requested
+				requested = true
+			}
```
[the remark this answers is not in the document]
Comment (zelig) [reply swarm/network/fetcher.go:139]: but that would be incorrect, see comment

File: swarm/network/fetcher.go [hunk 24]
```diff
@@ -0,0 +1,239 @@
...
+			// peer we requested from is gone. fall back to another
+			// and remove the peer from the peers map
+		case id := <-gone:
+			log.Debug("peer gone", "peer id", id.String(), "request addr", f.addr)
+			peers.Delete(id.String())
+			doRequest = true
+
+		// search timeout: too much time passed since the last request,
+		// extend the search to a new peer if we can find one
+		case <-waitC:
```
[the remark this answers is not in the document]
Comment (zelig) [reply swarm/network/fetcher.go:150]: true

File: swarm/network/fetcher.go [hunk 25]
```diff
@@ -0,0 +1,239 @@
...
+	var quit chan struct{}
+	var err error
+
+	req := &Request{
+		Addr:        f.addr,
+		SkipCheck:   f.skipCheck,
+		PeersToSkip: peersToSkip,
+	}
+	// iterate over known sources
+	for i = 0; i < len(sources); i++ {
```
[the remark this answers is not in the document]
Comment (zelig) [reply swarm/network/fetcher.go]: could but we only need index anyhow

File: swarm/network/fetcher.go [hunk 26]
```diff
@@ -0,0 +1,239 @@
...
+		req.Source = sources[i]
+		sourceID, quit, err = f.request(ctx, req)
+		if err == nil {
+			// remove the peer from known sources
+			// Note: we can modify the source although we are looping on it, because we break from the loop immediately
+			sources = append(sources[:i], sources[i+1:]...)
+			break
```
[the remark this answers is not in the document]
Comment (zelig) [reply swarm/network/fetcher.go]: ooops nice

File: swarm/network/fetcher_test.go [hunk 27]
```diff
@@ -0,0 +1,148 @@
...
+
+	select {
+	case request := <-requester.requestC:
+		for _, p := range peers {
+			if _, ok := request.PeersToSkip.Load(p); !ok {
+				t.Fatalf("request.peersToSkip misses peer")
+			}
+		}
+		// wait for the source peer to be added to peersToSkip
+		time.Sleep(100 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (zelig) [reply swarm/network/fetcher_test.go]: sync.map is concurrency safe but we still need to wait in the test for the other thread to change the contents

File: swarm/network/stream/common_test.go (hunk 1 above)
Comment (zelig) [reply swarm/network/stream/common_test.go:104]: possibly y

File: swarm/network/stream/delivery.go (hunk 2 above)
Comment (zelig) [reply swarm/network/stream/delivery.go:45]: ok not sure in this pr though

File: swarm/network/stream/messages.go (hunk 5 above)
Comment (zelig) [reply swarm/network/stream/messages.go]: no, explicit cancel. but i will think it through, this part is likely buggy

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 0b65a9bd53d75deb49637d3668f8054ec65c3c53
  * Commits must be in the following format: **%{scope}: %{description}**

 * Commit: 9001a9c496474ebb61039801b1b540e4f540a6da
  * Commits must be in the following format: **%{scope}: %{description}**

 * Commit: 10ea10f7cf69209ba85894b1dc03232cea3297d1
  * Commits must be in the following format: **%{scope}: %{description}**

 * Commit: 0a578d756d6a02add72bcd05e9cb73b3b4d6e450
  * Your subject line is longer than 72 characters

Guidelines are available at https://github.com/ethersphere/go-ethereum

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 0b65a9bd53d75deb49637d3668f8054ec65c3c53
  * Commits must be in the following format: **%{scope}: %{description}**

 * Commit: 9001a9c496474ebb61039801b1b540e4f540a6da
  * Commits must be in the following format: **%{scope}: %{description}**

 * Commit: 10ea10f7cf69209ba85894b1dc03232cea3297d1
  * Commits must be in the following format: **%{scope}: %{description}**

 * Commit: 0a578d756d6a02add72bcd05e9cb73b3b4d6e450
  * Your subject line is longer than 72 characters

Guidelines are available at https://github.com/ethersphere/go-ethereum

---
This message was auto-generated by https://gitcop.com

**Comment by nonsense** [comment]
I am trying to rebase this on top of master, and resolve the conflicts as a start. We can move on from there, once this is done.

**Pull request closed without merging**